pkg/storage/bloom/v1/bounds.go

```diff
@@ -0,0 +1,146 @@
+package v1
+
+import (
+	"hash"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/util/encoding"
+)
+
+type BoundsCheck uint8
+
+const (
+	Before BoundsCheck = iota
+	Overlap
+	After
+)
+
+type FingerprintBounds struct {
+	Min, Max model.Fingerprint
+}
+
+func NewBounds(min, max model.Fingerprint) FingerprintBounds {
+	return FingerprintBounds{Min: min, Max: max}
+}
+
+func (b FingerprintBounds) Hash(h hash.Hash32) error {
+	var enc encoding.Encbuf
+	enc.PutBE64(uint64(b.Min))
+	enc.PutBE64(uint64(b.Max))
+	_, err := h.Write(enc.Get())
+	return errors.Wrap(err, "writing OwnershipRange")
+}
+
+func (b FingerprintBounds) String() string {
+	return b.Min.String() + "-" + b.Max.String()
+}
+
+// Cmp returns the fingerprint's position relative to the bounds
+func (b FingerprintBounds) Cmp(fp model.Fingerprint) BoundsCheck {
+	if fp < b.Min {
+		return Before
+	} else if fp > b.Max {
+		return After
+	}
+	return Overlap
+}
+
+func (b FingerprintBounds) Overlaps(target FingerprintBounds) bool {
+	return b.Cmp(target.Min) != After && b.Cmp(target.Max) != Before
+}
+
+// Clip returns a new fingerprint bounds clipped to the target bounds or nil if there is no overlap
+func (b FingerprintBounds) Clip(target FingerprintBounds) *FingerprintBounds {
+	if !b.Overlaps(target) {
+		return nil
+	}
+
+	return &FingerprintBounds{
+		Min: max(b.Min, target.Min),
+		Max: min(b.Max, target.Max),
+	}
+}
+
+// Slice returns a new fingerprint bounds clipped to the target bounds or nil if there is no overlap
+func (b FingerprintBounds) Slice(min, max model.Fingerprint) *FingerprintBounds {
+	return b.Clip(FingerprintBounds{Min: min, Max: max})
+}
+
+// Returns whether the fingerprint is fully within the target bounds
+func (b FingerprintBounds) Within(target FingerprintBounds) bool {
+	return b.Min >= target.Min && b.Max <= target.Max
+}
+
+// Intersection returns the intersection of the two bounds
+func (b FingerprintBounds) Intersection(target FingerprintBounds) *FingerprintBounds {
```

Comment:
The body is the same as Clip. Make Intersection call Clip

Reply:
Haha, thanks. Going to prefer intersection but good catch